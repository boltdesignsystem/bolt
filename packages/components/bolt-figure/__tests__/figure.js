import {
  render,
  renderString,
  stopServer,
  html,
} from '../../../testing/testing-helpers';
const { media } = require('./figure.data');

describe('figure', () => {
  let page;

  afterAll(async () => {
    await stopServer();
    await page.close();
  });

  beforeEach(async () => {
    await page.evaluate(() => {
      document.body.innerHTML = '';
    });
  });

  beforeAll(async () => {
    page = await global.__BROWSER__.newPage();
    await page.goto('http://127.0.0.1:4444/', {
      timeout: 0,
    });
  });

  test("A Figures without any `default` slotted content won't render a <figcaption>", async function() {
    const renderedFigureHTML = await page.evaluate(async () => {
      document.body.insertAdjacentHTML(
        'beforeend',
        '<bolt-figure><img slot="media" src="/fixtures/landscape-16x9-mountains.jpg"></bolt-figure>',
      );
      const figure = document.querySelector('bolt-figure');
      await customElements.whenDefined('bolt-figure');
      await figure.updateComplete;

      // add a large border to the <bolt-figure> so an empty figcaption would better show up on the VRT
      figure.style.border = '4px solid green';

      return figure.renderRoot.innerHTML;
    });

    const renderedInnerHTML = await html('<div></div>');
    renderedInnerHTML.innerHTML = renderedFigureHTML;

    expect(
      renderedInnerHTML
        .querySelector('figure')
        .classList.contains('c-bolt-figure'),
    ).toBe(true);

    expect(renderedInnerHTML.querySelector('.c-bolt-figure__caption')).toBe(
      null,
    );

    expect(renderedInnerHTML).toMatchSnapshot();
  });

  test('Figures renders slotted `default` and `media` content', async function() {
    const renderedFigureHTML = await page.evaluate(async () => {
      document.body.insertAdjacentHTML(
        'beforeend',
        '<bolt-figure><img slot="media" src="/fixtures/landscape-16x9-mountains.jpg">Caption text goes here</bolt-figure>',
      );
      const figure = document.querySelector('bolt-figure');
      await customElements.whenDefined('bolt-figure');
      await figure.updateComplete;

      // add a large border to the <bolt-figure> so an empty figcaption would better show up on the VRT
      figure.style.border = '4px solid green';
      return figure.renderRoot.innerHTML;
    });

    const renderedHTML = await html(`<div></div>`);
    renderedHTML.innerHTML = renderedFigureHTML;

    expect(
      renderedHTML.querySelector('figure').classList.contains('c-bolt-figure'),
    ).toBe(true);

    expect(
      renderedHTML
        .querySelector('figcaption')
        .classList.contains('c-bolt-figure__caption'),
    ).toBe(true);

    expect(renderedHTML).toMatchSnapshot();
  });

  Object.keys(media).forEach(async item => {
    test(`figure with ${item}`, async () => {
      const data = JSON.stringify(media[item]);
      const results = await renderString(`
        {% include '@bolt-components-figure/figure.twig' with {
          media: {
            content: include('@bolt-components-${item}/${item}.twig', ${data})
          },
          caption: 'Figure with ${item}.'
         } %}
      `);
      expect(results.ok).toBe(true);
      expect(results.html).toMatchSnapshot();
    });

    test(`figure with deprecated "${item}" prop still renders`, async () => {
      const results = await render('@bolt-components-figure/figure.twig', {
        media: {
          [item]: media[item],
        },
        caption: `Figure with ${item}.`,
      });
      expect(results.ok).toBe(true);
      expect(results.html).toMatchSnapshot();
    });
  });
});
