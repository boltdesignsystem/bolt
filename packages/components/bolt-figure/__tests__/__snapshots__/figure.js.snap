// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`figure A Figures without any \`default\` slotted content won't render a <figcaption> 1`] = `
<div
  style=""
>
  <!---->
  
      
  <figure
    class="c-bolt-figure"
    style=""
  >
    
        
            
    <div
      class="c-bolt-figure__media"
      style=""
    >
      
              
      <slot
        name="media"
        style=""
      />
      
            
    </div>
    
          
        
      
  </figure>
  
    
  <!---->
</div>
`;

exports[`figure Figures renders slotted \`default\` and \`media\` content 1`] = `
<div
  style=""
>
  <!---->
  
      
  <figure
    class="c-bolt-figure"
    style=""
  >
    
        
            
    <div
      class="c-bolt-figure__media"
      style=""
    >
      
              
      <slot
        name="media"
        style=""
      />
      
            
    </div>
    
          
        
            
    <figcaption
      class="c-bolt-figure__caption"
      style=""
    >
      
              
      <slot
        style=""
      />
      
            
    </figcaption>
    
          
      
  </figure>
  
    
  <!---->
</div>
`;

exports[`figure figure with deprecated "icon" prop still renders 1`] = `
<bolt-figure>
  <figure class="c-bolt-figure">
    <ssr-keep for="bolt-figure"
              class="c-bolt-figure__media"
    >
      <div slot="media">
        <bolt-icon name="add-open"
                   size="large"
                   aria-hidden="true"
        >
          <span class="c-bolt-icon c-bolt-icon--add-open c-bolt-icon--large">
            <svg class="c-bolt-icon__icon c-bolt-icon__icon--large"
                 xmlns="http://www.w3.org/2000/svg"
                 viewbox="0 0 24 24"
            >
              <g fill="none"
                 fill-rule="evenodd"
              >
                <path>
                </path>
                <g fill="var(--bolt-theme-icon, currentColor)">
                  <path d="M12 21a9 9 0 110-18 9 9 0 010 18m0-20a11 11 0 100 22 11 11 0 000-22M16 11h-3V8c0-.5-.4-1-1-1a1 1 0 00-1 1v3H8a1 1 0 00-1 1c0 .6.5 1 1 1h3v3c0 .6.4 1 1 1s1-.4 1-1v-3h3c.6 0 1-.4 1-1s-.4-1-1-1">
                  </path>
                </g>
              </g>
            </svg>
          </span>
        </bolt-icon>
      </div>
    </ssr-keep>
    <figcaption class="c-bolt-figure__caption">
      <ssr-keep for="bolt-figure">
        Figure with icon.
      </ssr-keep>
    </figcaption>
  </figure>
</bolt-figure>
`;

exports[`figure figure with deprecated "image" prop still renders 1`] = `
<bolt-figure>
  <figure class="c-bolt-figure">
    <ssr-keep for="bolt-figure"
              class="c-bolt-figure__media"
    >
      <div slot="media">
        <bolt-image no-lazy
                    ratio="1151/638"
                    placeholder-color="hsl(233, 33%, 97%)"
                    placeholder-image="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw=="
                    style="background-color: hsl(233, 33%, 97%);"
                    src="/fixtures/landscape-16x9-mountains.jpg"
                    srcset="/fixtures/landscape-16x9-mountains-50.jpg 50w, /fixtures/landscape-16x9-mountains-100.jpg 100w, /fixtures/landscape-16x9-mountains-200.jpg 200w, /fixtures/landscape-16x9-mountains-320.jpg 320w, /fixtures/landscape-16x9-mountains-480.jpg 480w, /fixtures/landscape-16x9-mountains-640.jpg 640w, /fixtures/landscape-16x9-mountains-800.jpg 800w, /fixtures/landscape-16x9-mountains-1024.jpg 1024w, /fixtures/landscape-16x9-mountains.jpg 1151w"
        >
          <bolt-ratio ratio="1151/638">
            <ssr-keep for="bolt-ratio"
                      class="c-bolt-ratio"
                      style="padding-bottom: calc((638 / 1151) * 100%); --aspect-ratio: 1.8040752351097;"
            >
              <img class="c-bolt-image__image-placeholder"
                   alt
                   sizes="auto"
                   src="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw=="
              >
              <img class="c-bolt-image__image"
                   alt
                   srcset="/fixtures/landscape-16x9-mountains-50.jpg 50w, /fixtures/landscape-16x9-mountains-100.jpg 100w, /fixtures/landscape-16x9-mountains-200.jpg 200w, /fixtures/landscape-16x9-mountains-320.jpg 320w, /fixtures/landscape-16x9-mountains-480.jpg 480w, /fixtures/landscape-16x9-mountains-640.jpg 640w, /fixtures/landscape-16x9-mountains-800.jpg 800w, /fixtures/landscape-16x9-mountains-1024.jpg 1024w, /fixtures/landscape-16x9-mountains.jpg 1151w"
                   sizes="auto"
                   src="/fixtures/landscape-16x9-mountains.jpg"
              >
            </ssr-keep>
          </bolt-ratio>
        </bolt-image>
      </div>
    </ssr-keep>
    <figcaption class="c-bolt-figure__caption">
      <ssr-keep for="bolt-figure">
        Figure with image.
      </ssr-keep>
    </figcaption>
  </figure>
</bolt-figure>
`;

exports[`figure figure with deprecated "table" prop still renders 1`] = `
<bolt-figure>
  <figure class="c-bolt-figure">
    <ssr-keep for="bolt-figure"
              class="c-bolt-figure__media"
    >
      <div slot="media">
        <bolt-table format="regular">
          <table class="c-bolt-table">
            <thead class="c-bolt-table__head">
              <tr class="c-bolt-table__row">
                <td class="c-bolt-table__cell">
                </td>
                <th class="c-bolt-table__cell c-bolt-table__cell--header"
                    scope="col"
                >
                  Column 1
                </th>
                <th class="c-bolt-table__cell c-bolt-table__cell--header"
                    scope="col"
                >
                  Column 2
                </th>
                <th class="c-bolt-table__cell c-bolt-table__cell--header"
                    scope="col"
                >
                  Column 3
                </th>
              </tr>
            </thead>
            <tbody class="c-bolt-table__body">
              <tr class="c-bolt-table__row">
                <th class="c-bolt-table__cell c-bolt-table__cell--header"
                    scope="row"
                >
                  Row 1
                </th>
                <td class="c-bolt-table__cell c-bolt-table__cell--data">
                  R1C1
                </td>
                <td class="c-bolt-table__cell c-bolt-table__cell--data">
                  R1C2
                </td>
                <td class="c-bolt-table__cell c-bolt-table__cell--data">
                  R1C3
                </td>
              </tr>
              <tr class="c-bolt-table__row">
                <th class="c-bolt-table__cell c-bolt-table__cell--header"
                    scope="row"
                >
                  Row 2
                </th>
                <td class="c-bolt-table__cell c-bolt-table__cell--data">
                  R2C1
                </td>
                <td class="c-bolt-table__cell c-bolt-table__cell--data">
                  R2C2
                </td>
                <td class="c-bolt-table__cell c-bolt-table__cell--data">
                  R2C3
                </td>
              </tr>
              <tr class="c-bolt-table__row">
                <th class="c-bolt-table__cell c-bolt-table__cell--header"
                    scope="row"
                >
                  Row 3
                </th>
                <td class="c-bolt-table__cell c-bolt-table__cell--data">
                  R3C1
                </td>
                <td class="c-bolt-table__cell c-bolt-table__cell--data">
                  R3C2
                </td>
                <td class="c-bolt-table__cell c-bolt-table__cell--data">
                  R3C3
                </td>
              </tr>
            </tbody>
            <tfoot class="c-bolt-table__foot">
              <tr class="c-bolt-table__row">
                <th class="c-bolt-table__cell c-bolt-table__cell--header">
                  Footer
                </th>
                <td class="c-bolt-table__cell c-bolt-table__cell--data">
                  FC1
                </td>
                <td class="c-bolt-table__cell c-bolt-table__cell--data">
                  FC2
                </td>
                <td class="c-bolt-table__cell c-bolt-table__cell--data">
                  FC3
                </td>
              </tr>
            </tfoot>
          </table>
        </bolt-table>
      </div>
    </ssr-keep>
    <figcaption class="c-bolt-figure__caption">
      <ssr-keep for="bolt-figure">
        Figure with table.
      </ssr-keep>
    </figcaption>
  </figure>
</bolt-figure>
`;

exports[`figure figure with deprecated "video" prop still renders 1`] = `
<bolt-figure>
  <figure class="c-bolt-figure">
    <ssr-keep for="bolt-figure"
              class="c-bolt-figure__media"
    >
      <div slot="media">
        <bolt-ratio ratio="16/9">
          <ssr-keep for="bolt-ratio"
                    class="c-bolt-ratio"
                    style="padding-bottom: calc((9 / 16) * 100%); --aspect-ratio: 1.7777777777778;"
          >
            <bolt-video class="js-bolt-video-uuid--12345"
                        video-id="4892122320001"
                        account-id="1900410236"
                        show-meta
                        show-meta-title
                        player-id="r1CAdLzTW"
                        controls
                        share-description="Share This Video"
            >
            </bolt-video>
          </ssr-keep>
        </bolt-ratio>
      </div>
    </ssr-keep>
    <figcaption class="c-bolt-figure__caption">
      <ssr-keep for="bolt-figure">
        Figure with video.
      </ssr-keep>
    </figcaption>
  </figure>
</bolt-figure>
`;

exports[`figure figure with icon 1`] = `
<bolt-figure>
  <figure class="c-bolt-figure">
    <ssr-keep for="bolt-figure"
              class="c-bolt-figure__media"
    >
      <div slot="media">
        <bolt-icon name="add-open"
                   size="large"
                   aria-hidden="true"
        >
          <span class="c-bolt-icon c-bolt-icon--add-open c-bolt-icon--large">
            <svg class="c-bolt-icon__icon c-bolt-icon__icon--large"
                 xmlns="http://www.w3.org/2000/svg"
                 viewbox="0 0 24 24"
            >
              <g fill="none"
                 fill-rule="evenodd"
              >
                <path>
                </path>
                <g fill="var(--bolt-theme-icon, currentColor)">
                  <path d="M12 21a9 9 0 110-18 9 9 0 010 18m0-20a11 11 0 100 22 11 11 0 000-22M16 11h-3V8c0-.5-.4-1-1-1a1 1 0 00-1 1v3H8a1 1 0 00-1 1c0 .6.5 1 1 1h3v3c0 .6.4 1 1 1s1-.4 1-1v-3h3c.6 0 1-.4 1-1s-.4-1-1-1">
                  </path>
                </g>
              </g>
            </svg>
          </span>
        </bolt-icon>
      </div>
    </ssr-keep>
    <figcaption class="c-bolt-figure__caption">
      <ssr-keep for="bolt-figure">
        Figure with icon.
      </ssr-keep>
    </figcaption>
  </figure>
</bolt-figure>
`;

exports[`figure figure with image 1`] = `
<bolt-figure>
  <figure class="c-bolt-figure">
    <ssr-keep for="bolt-figure"
              class="c-bolt-figure__media"
    >
      <div slot="media">
        <bolt-image no-lazy
                    ratio="1151/638"
                    placeholder-color="hsl(233, 33%, 97%)"
                    placeholder-image="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw=="
                    style="background-color: hsl(233, 33%, 97%);"
                    src="/fixtures/landscape-16x9-mountains.jpg"
                    srcset="/fixtures/landscape-16x9-mountains-50.jpg 50w, /fixtures/landscape-16x9-mountains-100.jpg 100w, /fixtures/landscape-16x9-mountains-200.jpg 200w, /fixtures/landscape-16x9-mountains-320.jpg 320w, /fixtures/landscape-16x9-mountains-480.jpg 480w, /fixtures/landscape-16x9-mountains-640.jpg 640w, /fixtures/landscape-16x9-mountains-800.jpg 800w, /fixtures/landscape-16x9-mountains-1024.jpg 1024w, /fixtures/landscape-16x9-mountains.jpg 1151w"
        >
          <bolt-ratio ratio="1151/638">
            <ssr-keep for="bolt-ratio"
                      class="c-bolt-ratio"
                      style="padding-bottom: calc((638 / 1151) * 100%); --aspect-ratio: 1.8040752351097;"
            >
              <img class="c-bolt-image__image-placeholder"
                   alt
                   sizes="auto"
                   src="data:image/gif;base64,R0lGODlhAQABAAAAACH5BAEKAAEALAAAAAABAAEAAAICTAEAOw=="
              >
              <img class="c-bolt-image__image"
                   alt
                   srcset="/fixtures/landscape-16x9-mountains-50.jpg 50w, /fixtures/landscape-16x9-mountains-100.jpg 100w, /fixtures/landscape-16x9-mountains-200.jpg 200w, /fixtures/landscape-16x9-mountains-320.jpg 320w, /fixtures/landscape-16x9-mountains-480.jpg 480w, /fixtures/landscape-16x9-mountains-640.jpg 640w, /fixtures/landscape-16x9-mountains-800.jpg 800w, /fixtures/landscape-16x9-mountains-1024.jpg 1024w, /fixtures/landscape-16x9-mountains.jpg 1151w"
                   sizes="auto"
                   src="/fixtures/landscape-16x9-mountains.jpg"
              >
            </ssr-keep>
          </bolt-ratio>
        </bolt-image>
      </div>
    </ssr-keep>
    <figcaption class="c-bolt-figure__caption">
      <ssr-keep for="bolt-figure">
        Figure with image.
      </ssr-keep>
    </figcaption>
  </figure>
</bolt-figure>
`;

exports[`figure figure with table 1`] = `
<bolt-figure>
  <figure class="c-bolt-figure">
    <ssr-keep for="bolt-figure"
              class="c-bolt-figure__media"
    >
      <div slot="media">
        <bolt-table format="regular">
          <table class="c-bolt-table">
            <thead class="c-bolt-table__head">
              <tr class="c-bolt-table__row">
                <td class="c-bolt-table__cell">
                </td>
                <th class="c-bolt-table__cell c-bolt-table__cell--header"
                    scope="col"
                >
                  Column 1
                </th>
                <th class="c-bolt-table__cell c-bolt-table__cell--header"
                    scope="col"
                >
                  Column 2
                </th>
                <th class="c-bolt-table__cell c-bolt-table__cell--header"
                    scope="col"
                >
                  Column 3
                </th>
              </tr>
            </thead>
            <tbody class="c-bolt-table__body">
              <tr class="c-bolt-table__row">
                <th class="c-bolt-table__cell c-bolt-table__cell--header"
                    scope="row"
                >
                  Row 1
                </th>
                <td class="c-bolt-table__cell c-bolt-table__cell--data">
                  R1C1
                </td>
                <td class="c-bolt-table__cell c-bolt-table__cell--data">
                  R1C2
                </td>
                <td class="c-bolt-table__cell c-bolt-table__cell--data">
                  R1C3
                </td>
              </tr>
              <tr class="c-bolt-table__row">
                <th class="c-bolt-table__cell c-bolt-table__cell--header"
                    scope="row"
                >
                  Row 2
                </th>
                <td class="c-bolt-table__cell c-bolt-table__cell--data">
                  R2C1
                </td>
                <td class="c-bolt-table__cell c-bolt-table__cell--data">
                  R2C2
                </td>
                <td class="c-bolt-table__cell c-bolt-table__cell--data">
                  R2C3
                </td>
              </tr>
              <tr class="c-bolt-table__row">
                <th class="c-bolt-table__cell c-bolt-table__cell--header"
                    scope="row"
                >
                  Row 3
                </th>
                <td class="c-bolt-table__cell c-bolt-table__cell--data">
                  R3C1
                </td>
                <td class="c-bolt-table__cell c-bolt-table__cell--data">
                  R3C2
                </td>
                <td class="c-bolt-table__cell c-bolt-table__cell--data">
                  R3C3
                </td>
              </tr>
            </tbody>
            <tfoot class="c-bolt-table__foot">
              <tr class="c-bolt-table__row">
                <th class="c-bolt-table__cell c-bolt-table__cell--header">
                  Footer
                </th>
                <td class="c-bolt-table__cell c-bolt-table__cell--data">
                  FC1
                </td>
                <td class="c-bolt-table__cell c-bolt-table__cell--data">
                  FC2
                </td>
                <td class="c-bolt-table__cell c-bolt-table__cell--data">
                  FC3
                </td>
              </tr>
            </tfoot>
          </table>
        </bolt-table>
      </div>
    </ssr-keep>
    <figcaption class="c-bolt-figure__caption">
      <ssr-keep for="bolt-figure">
        Figure with table.
      </ssr-keep>
    </figcaption>
  </figure>
</bolt-figure>
`;

exports[`figure figure with video 1`] = `
<bolt-figure>
  <figure class="c-bolt-figure">
    <ssr-keep for="bolt-figure"
              class="c-bolt-figure__media"
    >
      <div slot="media">
        <bolt-ratio ratio="16/9">
          <ssr-keep for="bolt-ratio"
                    class="c-bolt-ratio"
                    style="padding-bottom: calc((9 / 16) * 100%); --aspect-ratio: 1.7777777777778;"
          >
            <bolt-video class="js-bolt-video-uuid--12345"
                        video-id="4892122320001"
                        account-id="1900410236"
                        show-meta
                        show-meta-title
                        player-id="r1CAdLzTW"
                        controls
                        share-description="Share This Video"
            >
            </bolt-video>
          </ssr-keep>
        </bolt-ratio>
      </div>
    </ssr-keep>
    <figcaption class="c-bolt-figure__caption">
      <ssr-keep for="bolt-figure">
        Figure with video.
      </ssr-keep>
    </figcaption>
  </figure>
</bolt-figure>
`;
